contracts/zaps/ZapCurveTokenToPrincipalToken.sol

```diff
@@ -0,0 +1,458 @@
+    /// their respective lp tokens and then immediately swaps them for
+    /// principal tokens, greatly simplifying the purchasing of principal
+    /// tokens
+    /// @param _info See ZapInInfo struct
+    /// @param _zap See ZapCurveLpIn struct - This is the "main" or parent zap
+    /// which produces the lp token necessary to swap for the principal token
+    /// @param _childZap See ZapCurveLpIn - This is used only in cases where
+    /// the "main" or "parent" zap itself is composed of another curve lp token
+    /// which can be accessed more readily via another swap via curve
+    function zapIn(
+        ZapInInfo memory _info,
+        ZapCurveLpIn memory _zap,
+        ZapCurveLpIn memory _childZap,
+        PermitData[] memory _permitData
+    )
+        external
+        payable
+        notFrozen
+        preApproval(_permitData)
+        returns (uint256 ptAmount)
+    {
+        // Instantiation of the context amount container which is used to track
+        // amounts to be swapped in the final curve zap.
+        uint256[3] memory ctx;
+
+        // Only execute the childZap if it is necessary
+        if (_info.needsChildZap) {
+            uint256 _amount = _zapCurveLpIn(
+                _childZap,
+                // The context array is unnecessary for the childZap and so we
+                // can just put a dud array in place of it
+                [uint256(0), uint256(0), uint256(0)]
+            );
+            // When a childZap happens, we add the amount of lpTokens gathered
+            // from it to the relevant root index of the "main" zap
+            ctx[_childZap.parentIdx] += _amount;
+        }
+
+        // Swap an amount of "root" tokens on curve for the lp token that is
+        // used to then purchase the principal token
+        uint256 baseTokenAmount = _zapCurveLpIn(_zap, ctx);
+
+        // Purchase of "ptAmount" of principal tokens
+        ptAmount = _balancer.swap(
+            IVault.SingleSwap({
+                poolId: _info.balancerPoolId,
+                kind: IVault.SwapKind.GIVEN_IN,
+                assetIn: IAsset(address(_zap.lpToken)),
+                assetOut: _info.principalToken,
+                amount: baseTokenAmount,
+                userData: "0x00"
+            }),
+            IVault.FundManagement({
+                sender: address(this),
+                fromInternalBalance: false,
+                recipient: payable(_info.recipient),
+                toInternalBalance: false
+            }),
+            _info.minPtAmount,
+            _info.deadline
+        );
+    }
+
+    struct ZapCurveLpOut {
+        // Address of the curvePool for which an amount of lpTokens
+        // is swapped for an amount of single root tokens
+        address curvePool;
+        // The contract address of the curve pools lpToken
+        IERC20 lpToken;
+        // This is the index of the target root we are swapping for
+        int128 rootTokenIdx;
```

Comment:
How we are going to validate the index of the token that the user is going to get.

Reply:
We could do it by the `coins` method on the curve pool I mentioned in another comment above but may be more trouble than it's worth if func sig must be passed from off-chain